### PR TITLE
Fixing issue where ScriptingCore::evalString would always return fals…

### PR DESCRIPTION
**Pull request by DavidDeSimone**
…e, and not evaluate the string, due to an improper null check.

### CODE REVIEW

**Comment by DavidDeSimone** [comment]
Further background can be found in this issue: https://github.com/cocos2d/cocos2d-x/issues/16541

**Comment by minggo** [comment]
@pandamicro could you please take a look?

**Review by pandamicro — APPROVED** [review]

**Comment by minggo** [comment]
@DavidDeSimone could you please add a test case for this issue?

File: cocos/scripting/js-bindings/manual/ScriptingCore.cpp [hunk 1]
```diff
@@ -520,8 +520,8 @@ void ScriptingCore::string_report(JS::HandleValue val) {
 bool ScriptingCore::evalString(const char *string, JS::MutableHandleValue outVal, const char *filename, JSContext* cx, JS::HandleObject global)
 {
     JSAutoCompartment ac(cx, global);
-    JS::PersistentRootedScript script(cx);
-    if (script == nullptr) {
+    JS::RootedScript script(cx);
+    if (&script == nullptr) {
```
Comment (ricardoquesada) [inline cocos/scripting/js-bindings/manual/ScriptingCore.cpp]: I think that `script` can't be `nullptr` with the new change as well.
And why did you change `PersistentRootedScript` to `RootedScript`? thanks.

Comment (DavidDeSimone) [reply cocos/scripting/js-bindings/manual/ScriptingCore.cpp]: @ricardoquesada I believe you're correct. I don't believe that if check is necessary. It seems likely that if check is a hold over from when we created a PersistentRootedObject on the heap here. It makes sense to do that check if we are doing a dynamic allocation, but since we are creating a Rooted<T> on the stack, I don't believe we need to check the value of operator& on the Rooted<T>. I don't believe we if check other Rooted<T>s we create on the stack in other places in code, so it makes sense to not have that if check here. 

w.r.t the PersistentRootedScript, In the issue I wrote 

> Also on a slightly related note, I am unsure on why we are using a JS::PersistentRootedScript here over a JS::RootedScript. The script in this use case has a definite lifetime, and is not cached or stored globally, so I see no point for using a PersistentRooted here.

The spidermonkey API suggests that PersistentRooted<T>s should be used for objects with dynamic lifetime that could change based on runtime parameters. Since the lifetime of the script object here is known at compile time, it seems more proper to use a JS::Rooted<T> instead of a PersistentRooted<T>.

**Comment by DavidDeSimone** [comment]
@minggo I will write up some test cases either tonight or tomorrow and add them to the PR.

**Comment by DavidDeSimone** [comment]
Updated the PR based on feedback from @ricardoquesada

As for the tests, it seems that there isn't a binding for evalString, and because of that I'm not sure of the best way to add a test to the js testing framework for it. As far as I can see, the tests in tests/js-tests only test functionality available directly from the JS, while this is a c++ only function. 

I could add a binding for evalString in scripting core, and write a test that uses it, but I'm curious what other people think.

**Comment by minggo** [comment]
@ricardoquesada any update about this PR since @DavidDeSimone updated the codes according to your comment.

**Comment by DavidDeSimone** [comment]
@minggo @pandamicro It looks like @pandamicro fixed this issue in https://github.com/cocos2d/cocos2d-x/pull/16751, should we just close this PR now as duplicate?

**Comment by minggo** [comment]
@DavidDeSimone thanks for the information
